ambry-store/src/main/java/com/github/ambry/store/BlobStoreCompactor.java

```diff
@@ -1295,4 +998,533 @@ private void endCompaction() {
+      //                     | Undelete(f)       | true                                         |
+      // --------------------+-------------------+----------------------------------------------|
+      // Delete(c)           | Put(f)            | Exception                                    |
+      //                     | Delete(f)         | c==f?true:false                              |
+      //                     | Undelete(f)       | false                                        |
+      // --------------------+-------------------+----------------------------------------------|
+      // Undelete(c)         | Put(f)            | Exception                                    |
+      //                     | Delete(f)         | false                                        |
+      //                     | Undelete(f)       | c==f&&!isExpired(Uc)?true:false              |
+      // ----------------------------------------------------------------------------------------
+      List<IndexEntry> validEntries = new ArrayList<>();
+      Iterator<IndexEntry> iterator = indexSegment.getIterator();
+      StoreKey lastKey = null;
+      IndexValue lastFinalState = null;
+      while (iterator.hasNext()) {
+        IndexEntry entry = iterator.next();
+        StoreKey currentKey = entry.getKey();
+        IndexValue currentValue = entry.getValue();
+        IndexValue currentFinalState;
+
+        if (lastKey == null) {
+          // save a token for restart (the key gets ignored but is required to be non null for construction)
+          StoreFindToken safeToken =
+              new StoreFindToken(currentKey, indexSegment.getStartOffset(), sessionId, incarnationId);
+          compactionLog.setSafeToken(safeToken);
+          logger.debug("Set safe token for compaction in {} to {}", storeId, safeToken);
+        }
+        if (currentKey.equals(lastKey)) {
+          currentFinalState = lastFinalState;
+        } else {
+          lastKey = currentKey;
+          lastFinalState = currentFinalState = srcIndex.findKey(currentKey);
+        }
+
+        if (currentValue.isUndelete()) {
+          if (currentFinalState.isPut()) {
+            throw new IllegalStateException(
+                "Undelete's final state can't be put for key" + currentKey + " in store " + dataDir);
+          }
+          if (currentFinalState.isUndelete() && currentFinalState.getLifeVersion() == currentValue.getLifeVersion()
+              && !srcIndex.isExpired(currentValue)) {
+            validEntries.add(entry);
+          }
+        } else if (currentValue.isDelete()) {
+          if (currentFinalState.isPut()) {
+            throw new IllegalStateException(
+                "Delete's final state can't be put for key" + currentKey + " in store " + dataDir);
+          }
+          if (currentFinalState.isDelete() && currentFinalState.getLifeVersion() == currentValue.getLifeVersion()) {
```

Comment:
why we don't check retention time here?

Reply:
We don't have to. The current value is delete, and it's only valid when it's the last delete for this blob. eg. P -> D0 -> U1 -> D1. If D0 is the current value, then it's not valid, since D1 is the last delete. eg. P -> D0 -> U1 -> D1 -> U2. If D1 is the current value, then it's not valid, since final state is not a delete.